Update to 1.20.6 and fix Comet

mms0316 opened the pull request:
- Several renames had to be done as Spigot/Paper API changed names to actual Minecraft names
- PotionType.UNCRAFTABLE no longer exists, affecting Sake and its (non-)interaction with Potion Sachel.
  - To fix this, a new namespace key was used to be able to identify Sake
  - I did not want to use the existing namespace HoloItems because this would force consumables (like Sake) to start implementing the Consumable interface. Note that it's for this same reason that Secret Brew potions can't be consumed, only thrown, which looks to be an oversight bug.
- Comet is reenabled and use of Armor Stand were changed to ItemDisplay
- I did not try to update to 1.21.1 as a lot of Spigot/Paper API was removed. Do tell me if this should be done (in another PR), and I might actually have to do it for Protocol Lib, since Entity Type IDs looked to be different in each Minecraft version in the documentation
- Update to Java 21 (required for 1.20.5+)

mms0316 committed:
Update to 1.20.6 - easy renames

mms0316 committed:
Update to 1.20.6 - non-trivial renames

mms0316 committed:
Update to 1.20.6 - handle removal of PotionType.UNCRAFTABLE
- Used by Sake, and PotionSatchel had code to specifically ignore PotionType.UNCRAFTABLE
- Changed to a Mundane Potion, which may have unintended consequences (by being able to be converted in other recipes)
- Added new namespace "id" to be able to specifically recognize Sake. If the already existing namespace "key" was used, it would need more changes (such as inheriting Consumable in order to be able to be right-clicked)

mms0316 committed:
fix Comet
- target.getCategory()  throws exceptions in > 1.20.5, and emits a warning to use Tags

mms0316 committed:
Update Java to 21 for Paper API 1.20.5+

mms0316 committed:
Update dependencies

mms0316 committed:
Update workflow to Java 21

mms0316 committed:
Remove unneeded code for Comet

mms0316 committed:
Change use of Armor Stand to ItemDisplay

dlee13 commented on src/main/java/com/klin/holoItems/collections/gen0/suiseiCollection/items/Comet.java:
```diff
@@ -108,8 +106,7 @@ public void ability(PlayerInteractEvent event, Action action){
                     entity -> (entity != player &&
-                            entity instanceof LivingEntity && !(entity instanceof ArmorStand) &&
-                            !targets.contains(entity))); // TODO: remove this, since Set<> should be enough
+                            entity instanceof LivingEntity && !(entity instanceof ArmorStand)));
```
Wouldn't you get the same RayTraceResult multiple times if you change this?

mms0316 replied:
You're right, I got misled by usage of Set, but it's really needed to filter during ray tracing. It's reverted.

mms0316 committed:
Revert attempt to simplify ray tracing in Comet
- It was misleading due to usage of Set. The readded code is needed as a filter to rayTrace in order for Piercing to work

mms0316 committed:
Change positioning closer to old Comet

mms0316 commented:
One thing that I really didn't bring was the sinusoidal movement on X axis (old Comet moved like _.-._.-._), and if you compare, old Comet movement was smoother (less choppy). This is very noticeable when using /tick rate below 20

I haven't figured out if there's a way to add interpolation when the axe moves forward

mms0316 committed:
Remove Comet from being shown in minimaps

dlee13 commented on src/main/java/com/klin/holoItems/collections/gen0/suiseiCollection/items/Comet.java:
```diff
@@ -73,113 +76,152 @@ public void registerRecipes(){
+
+        Location axeLocation = location.clone().add(handOffset).add(backwardOffset).add(downwardOffset);
+
+        ItemDisplay axeDisplay = world.spawn(axeLocation, ItemDisplay.class);
+        axeDisplay.setInvisible(true); // Remove mark in minimaps
+        axeDisplay.setItemStack(item);
+        axeDisplay.setViewRange((float)maxDistance);
```
```suggestion
        ItemDisplay axeDisplay = world.spawn(axeLocation, ItemDisplay.class);
        axeDisplay.setInvisible(true); // Remove mark in minimaps
        axeDisplay.setPersistent(false);
        axeDisplay.setItemStack(item);
        axeDisplay.setViewRange((float)maxDistance);
```
We could've avoided the whole detached armor stands problem if the original implementation did this too... But I never really looked closely at Comet's code until now.

dlee13 commented:
> Also, old Comet movement was smoother (less choppy). This is very noticeable when using /tick rate below 20
> 
> I haven't figured out if there's a way to add interpolation when the axe moves forward

I believe you're supposed to set the translation once and then the display entity will move every tick with interpolation automatically applied in the client side.

mms0316 committed:
Remove axe ItemDisplay if chunk unloads

mms0316 committed:
Fix Comet initial positioning
- Fixed initial position being different depending on player's angle, by using Translation
- Moved initialization to a Consumer, so it is set before a game tick. This fixes entity dot in minimap in the first tick
- Fixed offhand slant

mms0316 committed:
Fix Comet animation
- Animation was choppy, due to use of teleport(). Fixed by using Translation.
- Note: changed raytracing to use the normalized vector instead of x3 speed vector. This didn't change anything in testing.

mms0316 commented:
Looks good to me now
